### PR TITLE
mgr/dashboard: Adding legends and fixing queries

### PR DESCRIPTION
**Pull request by cloudbehl**
1. Fixing Prometheus Queries.
2. Fixing step count for prometheus queries(which directly impact in number of data points that will be used to provide granularity).
3. Adding Legend on the Left.
     - If we have max capacity available we will show as `used capacity`
     - if we have two legends we are naming them. Example `OPS` and `IPS`
     - If we have single graph without max, we will show as `Recovery Throughput`
5. Fixing tension and border with for line charts to look cleaner.
6. Removing Animation to make graph movement smooth.
7. Improving the date format in legend `DD/MM/YYYY - HH:MM:SS`.
8. Adding units n legends next to number.
9. Fixing timing issue related to graph plotting which was leading to plot 0 as last value and legends were also showing 0 instead of number.
10. Adding color to the legends.
11. fixing name for graphs. 
12. Restricting the timer to max 24 hours rather than 7 days.

Fixes: https://tracker.ceph.com/issues/62369

**New:**

[Screencast from 2023-08-10 18-35-49.webm](https://github.com/ceph/ceph/assets/6695156/85afd426-2e5a-4394-9b54-722b9ac76901)

**OLD:**

https://github.com/ceph/ceph/assets/6695156/18dc0494-47e2-49b5-b248-d1932ef4bce9

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by pcuzner** [comment]
Looks better and having the video examples here is a great idea! However, maybe it's me but I still find a number of anomalies;

Nitpick - but "Cluster utilization" should be "Cluster Utilization"

Are you pinning the tooltip to the left yaxis to avoid the truncation issue? If so it seems a bit of a weird UX - not sure it aligns to https://www.patternfly.org/charts/line-chart either.

I'm confused by the labels of the IOPS chart. What are OPS and IPS..why these names? Surely they should be read and write? I'm also assuming they are sums of the pool_r and pool_wr metrics and not physical IOPS from the OSDs (so they are more meaningful against the client throughput chart)

Is there any reason why the colors are not in the legend? If there were it's easier to get an appreciation for the chart without  mouseover.

What's the relationship between the IOPS and client throughput? For example, how does it show a IOPS of >2000, when the client throughput only shows 33KB - this implies a workload of 16byte IO size?

it's unclear whether the graph tooltip shows both values at any point on the graph...it should

I'd recommend 
- changing the "Used Capacity" title to "Raw Capacity" to avoid any misunderstandings (i.e. it's not used/usable capacity you're showing...right?)
- prefixing the Latency title with OSD, so there are no misunderstandings (this isn't related to clients!) I also assume the query is an average?
- don't use 3 decimal places for the latency - fix it to 2 at most. In the demo, it goes from 0 to 1 to 3 decimal places.

Also, are you sure you want to show a timepicker that reaches back 7 days on the landing page? How does that perform on a large cluster? I would have thought the landing page should focus on the 'now', restricting the view to shorter intervals to exploit the prometheus cache (15min, 30min, 1hr, 3hrs for example). Longer intervals are available in the timepickers on the embeeded grafana dashboards.

**Comment by cloudbehl** [comment]
> Looks better and having the video examples here is a great idea! However, maybe it's me but I still find a number of anomalies;

Thanks, I agree that we have some anomalies and we are trying to improve it.

> 
> Nitpick - but "Cluster utilization" should be "Cluster Utilization"
Thanks, Done!
> 
> Are you pinning the tooltip to the left yaxis to avoid the truncation issue? If so it seems a bit of a weird UX - not sure it aligns to https://www.patternfly.org/charts/line-chart either.

Yes this is a to avoid the truncation issue as this Issue lies with the library(chartJS), we need to create a custom tooltip and to do it properly. 

> 
> I'm confused by the labels of the IOPS chart. What are OPS and IPS..why these names? Surely they should be read and write? I'm also assuming they are sums of the pool_r and pool_wr metrics and not physical IOPS from the OSDs (so they are more meaningful against the client throughput chart)

You are correct, the OPS/IPS doesn't make sense and this is Reads and Writes for the OSD and the units should  be in MiB, KiB etc. We are using `ceph_osd_op_w_in_bytes` and `ceph_osd_op_r_out_bytes` metric. I'm also changing the name of graph to Cluster I/O which is what we use in grafana. 

> 
> Is there any reason why the colors are not in the legend? If there were it's easier to get an appreciation for the chart without mouseover.

I'll add this

> 
> What's the relationship between the IOPS and client throughput? For example, how does it show a IOPS of >2000, when the client throughput only shows 33KB - this implies a workload of 16byte IO size?

As mentioned above, the units were incorrect for Reads and Writes for IOPS and I have fixed it. I'm not sure how they related in terms of time series data what I can see IOPS are OSD metrics and Client throughtput in Pool metric.

```
**IOPS**
  WRITEIOPS = 'sum(rate(ceph_osd_op_w_in_bytes[1m]))',
  READIOPS = 'sum(rate(ceph_osd_op_r_out_bytes[1m]))',
```

```
**Client throughput:**
  READCLIENTTHROUGHPUT = 'sum(rate(ceph_pool_rd_bytes[1m]))',
  WRITECLIENTTHROUGHPUT = 'sum(rate(ceph_pool_wr_bytes[1m]))',

```
> 
> it's unclear whether the graph tooltip shows both values at any point on the graph...it should

Done added both values at any point.

> 
> I'd recommend
> 
> * changing the "Used Capacity" title to "Raw Capacity" to avoid any misunderstandings (i.e. it's not used/usable capacity you're showing...right?)

Do we get usable capacity? Also would it be fine if we call it `Used Capacity(Raw)`?

> * prefixing the Latency title with OSD, so there are no misunderstandings (this isn't related to clients!) I also assume the query is an average?

Done

> * don't use 3 decimal places for the latency - fix it to 2 at most. In the demo, it goes from 0 to 1 to 3 decimal places.

Done

> 
> Also, are you sure you want to show a timepicker that reaches back 7 days on the landing page? How does that perform on a large cluster? I would have thought the landing page should focus on the 'now', restricting the view to shorter intervals to exploit the prometheus cache (15min, 30min, 1hr, 3hrs for example). Longer intervals are available in the timepickers on the embeeded grafana dashboards.

I have not given a thought, maybe we could do that. Performance wise we don't have issue as Prometheus it managing it via step count. 


```
**Queries used in graphs:**
  USEDCAPACITY = 'ceph_cluster_total_used_bytes',
  WRITEIOPS = 'sum(rate(ceph_osd_op_w_in_bytes[1m]))',
  READIOPS = 'sum(rate(ceph_osd_op_r_out_bytes[1m]))',
  READLATENCY = 'avg_over_time(ceph_osd_apply_latency_ms[1m])',
  WRITELATENCY = 'avg_over_time(ceph_osd_commit_latency_ms[1m])',
  READCLIENTTHROUGHPUT = 'sum(rate(ceph_pool_rd_bytes[1m]))',
  WRITECLIENTTHROUGHPUT = 'sum(rate(ceph_pool_wr_bytes[1m]))',
  RECOVERYBYTES = 'sum(rate(ceph_osd_recovery_bytes[1m]))'
```

**Comment by pcuzner** [comment]
Nice work - that looks a lot better, and thanks for showing the queries! My only reservation is the IOPS queries - they focus on backend since they're using the ceph_osd* metrics. I would tend to show a sum of the ceph_pool_rd and ceph_pool_rw since that is the load that is user generated and you'd be concerned about, not the underlying IO to the drives (that's more of an under-the-covers thing IMO).

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-details/rgw-user-details.component.ts [hunk 1]
```diff
@@ -98,10 +97,6 @@ export class RgwUserDetailsComponent implements OnChanges, OnInit {
     }
   }
 
-  updateKeysSelection(selection: CdTableSelection) {
-    this.keysSelection = selection;
-  }
-
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-details/rgw-user-details.component.ts:104]: why's this getting removed here?

**Review by nizamial09 — COMMENTED** [review]
Awesome work @cloudbehl 

I am seeing an undefined in the `Recovery Throughput` chart when there isn't anything
Screenshot from 2023-08-17 11-05-26

Charts look a bit crammed up to me! Can we increase the padding between each charts a little bit more?
Screenshot from 2023-08-17 11-05-47

**Comment by nizamial09** [comment]
the status is card is breaking a bit. not just by this PR but a series of PR went in on the landing page (Exposing of API and Telemetry URL). We'll need to invest some time into that too probably before getting these fixes into the stable branches.

Screenshot from 2023-08-17 11-43-26

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-details/rgw-user-details.component.ts (hunk 1 above)
Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-details/rgw-user-details.component.ts]: I think merge conflict created this, I'll remove.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.ts [hunk 2]
```diff
@@ -37,33 +37,37 @@ export class DashboardAreaChartComponent implements OnChanges, AfterViewInit {
...
       },
       {
         label: '',
         data: [],
-        tension: 0,
+        tension: 0.2,
         pointBackgroundColor: this.cssHelper.propertyValue('chart-color-orange'),
-        backgroundColor: this.cssHelper.propertyValue('chart-color-yellow'),
-        borderColor: this.cssHelper.propertyValue('chart-color-orange')
+        backgroundColor: this.cssHelper.propertyValue('chart-color-translucent-yellow'),
```
[the remark this answers is not in the document]
Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.ts:60]: Good find, I created the variable, looks like it got stashed.

**Comment by cloudbehl** [comment]
> the status is card is breaking a bit. not just by this PR but a series of PR went in on the landing page (Exposing of API and Telemetry URL). We'll need to invest some time into that too probably before getting these fixes into the stable branches.
> 
> Screenshot from 2023-08-17 11-43-26

I agree this needs to be fixed. For now maybe we can align alert to center, but it overall needs a better revamped a bit.

**Comment by cloudbehl** [comment]
jenkins test make check

**Comment by cloudbehl** [comment]
jenkins test dashboard

**Comment by cloudbehl** [comment]
jenkins test dashboard

**Comment by cloudbehl** [comment]
jenkins test dashboard

**Comment by cloudbehl** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test api

**Review by aaSharma14 — APPROVED** [review]
Tested Locally! LGTM

**Pull request merged**